lib/reservation_tracker.rb

```diff
@@ -0,0 +1,267 @@
+
+module Hotel
+  class ReservationTracker
+    attr_reader :rooms, :blocks, :reservations
+
+    def initialize(room_file = 'support/rooms.csv', block_file = 'support/blocks.csv', reservation_file = 'support/reservations.csv')
+      @rooms = load_rooms(room_file)
+      @blocks = load_blocks(block_file)
+      @reservations = load_reservations(reservation_file)
+    end
+
+    def load_rooms(filename)
+      rooms = []
+
+      CSV.read(filename, headers: true).each do |line|
+
+        input_data = {}
+        input_data[:room_num] = line[0].to_i
+        input_data[:rate] = line[1].to_i
+
+        room = Room.new(input_data)
+        rooms << room
+      end
+      return rooms
+    end
+
+    def load_blocks(filename)
+      blocks = []
+      block_data = CSV.open(filename, 'r', headers: true, header_converters: :symbol)
+
+      block_data.each do |raw_block|
+        start_date = Date.parse(raw_block[:start_date])
+        end_date = Date.parse(raw_block[:end_date])
+
+        requested_qty = raw_block[:party].to_i
+        block_qty_rooms = @rooms.take(requested_qty)
+        requested_dates = DateRange.new(start_date, end_date)
+
+        parsed_block = {
+          id: raw_block[:id].to_i,
+          party: block_qty_rooms,
+          date_range: requested_dates
+        }
+
+        block = Block.new(parsed_block)
+        blocks << block
+      end
+      return blocks
+    end
+
+    def load_reservations(filename)
+      reservations = []
+      reservation_data = CSV.open(filename, 'r', headers: true, header_converters: :symbol)
+
+      reservation_data.each do |raw_reservation|
+        block_id = find_block_id(raw_reservation[:block_id].to_i)
+        room = find_room(raw_reservation[:room].to_i)
+
+        start_date = Date.parse(raw_reservation[:start_date])
+        end_date = Date.parse(raw_reservation[:end_date])
+
+        range = DateRange.new(start_date, end_date)
+
+        parsed_reservation = {
+          id: raw_reservation[:id].to_i,
+          block_id: block_id,
+          room: room,
+          date_range: range
+        }
+
+        reservation = Reservation.new(parsed_reservation)
+        reservations << reservation
+      end
+      return reservations
+    end
+
+    def find_block_id(id)
+      check_id(id)
+      return @blocks.find { |block| block.id == id }
+    end
+
+    def find_room(room_num)
+      check_room_num(room_num)
+      return @rooms.find { |room| room.room_num == room_num }
+    end
+
+    def list_reservations_by_date(specified_date)
+      reservations_by_date = @reservations.find_all do |reservation|
+        date_range = reservation.date_range.get_range
+        date_range.include?(specified_date)
+      end
+      return reservations_by_date
+    end
+
+    def reservations_overlaps?(requested_dates)
+      matching_reservations = @reservations.find_all do |reservation|
+        date_range = reservation.date_range
+        reservation if date_range.overlaps?(requested_dates)
```

Comment:
I have several things to say here.

First, this is a great example of loose coupling. Rather than pulling out the dates and doing a bunch of math yourself, you call a method and let the date range take care of it. Right on!

Second, I would like to see an explicit return here (i.e. `return` instead of `matching_reservations =`, or end with `return matching_reservations`).

Third, I think you're somewhat misusing `find_all`. The block to `find_all` should produce a true or false value, whereas yours produces either the `reservation` (if the date range overlaps) or the `nil` (if it does not). This ends up working, but is not particularly clear. I would rewrite this loop as:

```ruby
matching_reservations = @reservations.find_all do |reservation|
  reservation.date_range.overlaps?(requested_dates)
end
```